src/Plugins/ComplexCore/src/ComplexCore/Filters/Algorithms/FindFeatureClustering.cpp

```diff
@@ -0,0 +1,297 @@
+  const auto zPoints = static_cast<usize>(boxDims[2] / boxRes[2]);
+
+  const usize totalPoints = xPoints * yPoints * zPoints;
+
+  const float32 stepSize = (maxDistance - minDistance) / numBins;
+  const float32 maxBoxDistance = sqrtf((boxDims[0] * boxDims[0]) + (boxDims[1] * boxDims[1]) + (boxDims[2] * boxDims[2]));
+  const auto currentNumBins = static_cast<usize>(ceil((maxBoxDistance - minDistance) / stepSize));
+
+  freq.resize(static_cast<size_t>(currentNumBins + 1));
+
+  std::random_device randomDevice;           // Will be used to obtain a seed for the random number engine
+  std::mt19937_64 generator(randomDevice()); // Standard mersenne_twister_engine seeded with rd()
+  std::mt19937::result_type seed = userSeedValue;
+  if(!useSeedFromUser)
+  {
+    seed = static_cast<std::mt19937::result_type>(std::chrono::steady_clock::now().time_since_epoch().count());
+  }
+  generator.seed(seed);
+  std::uniform_real_distribution<double> distribution(0.0, 1.0);
+
+  randomCentroids.resize(largeNumber * 3);
+
+  // Generating all of the random points and storing their coordinates in randomCentroids
+  for(usize i = 0; i < largeNumber; i++)
+  {
+    const auto featureOwnerIdx = static_cast<usize>(distribution(generator) * totalPoints);
+
+    const usize column = featureOwnerIdx % xPoints;
+    const usize row = (featureOwnerIdx / xPoints) % yPoints;
+    const usize plane = featureOwnerIdx / (xPoints * yPoints);
+
+    const auto xc = static_cast<float>(column * boxRes[0]);
+    const auto yc = static_cast<float>(row * boxRes[1]);
+    const auto zc = static_cast<float>(plane * boxRes[2]);
+
+    randomCentroids[3 * i] = xc;
+    randomCentroids[3 * i + 1] = yc;
+    randomCentroids[3 * i + 2] = zc;
+  }
+
+  distanceList.resize(largeNumber);
+
+  // Calculating all of the distances and storing them in the distance list
+  for(size_t i = 1; i < largeNumber; i++)
+  {
+    const float32 x = randomCentroids[3 * i];
+    const float32 y = randomCentroids[3 * i + 1];
+    const float32 z = randomCentroids[3 * i + 2];
+
+    for(size_t j = i + 1; j < largeNumber; j++)
+    {
+
+      const float32 xn = randomCentroids[3 * j];
+      const float32 yn = randomCentroids[3 * j + 1];
+      const float32 zn = randomCentroids[3 * j + 2];
+
+      float32 r = sqrtf((x - xn) * (x - xn) + (y - yn) * (y - yn) + (z - zn) * (z - zn));
+
+      distanceList[i].push_back(r);
+      distanceList[j].push_back(r);
+    }
+  }
+
+  // bin up the distance list
+  for(usize i = 0; i < largeNumber; i++)
+  {
+    for(const auto distance : distanceList[i])
+    {
+      const auto bin = static_cast<usize>((distance - minDistance) / stepSize);
+
+      if(distance < minDistance)
+      {
+        freq[0]++;
+      }
+      else
+      {
+        freq[bin + 1]++;
+      }
+    }
+  }
+
+  // Normalize the frequencies
+  for(size_t i = 0; i < currentNumBins + 1; i++)
+  {
+    freq[i] = freq[i] / (numDistances);
```

Comment:
```suggestion
    freq[i] /= (numDistances);
```
If you have already considered this and it didn't work, then just mark resolved (I know the assignment operators can be weird sometimes)